src/Uno.UWP/Globalization/ApplicationLanguages.cs

```diff
@@ -20,16 +20,64 @@ public static partial class ApplicationLanguages
 		static ApplicationLanguages()
 		{
 #if !NET461
-			if (ApplicationData.Current.LocalSettings.Values.TryGetValue(PrimaryLanguageOverrideSettingKey, out var savedValue)
-				&& savedValue is string stringSavedValue)
+			var appSpecificKey = GetAppSpecificSettingKey();
+			var hasLegacySetting = TryGetSettingFromKey(PrimaryLanguageOverrideSettingKey, out var legacySavedValue);
+			var hasAppSpecificSetting = TryGetSettingFromKey(appSpecificKey, out var appSpecificSavedValue);
+
+			// We have four cases:
+			// 1. hasAppSpecificSetting && hasLegacySetting
+			// 2. hasAppSpecificSetting && !hasLegacySetting
+			// 3. !hasAppSpecificSetting && hasLegacySetting
+			// 4. !hasAppSpecificSetting && !hasLegacySetting
+
+			// For 1 and 2, we will ignore the legacy setting and use the app specific setting.
+			// For 3, we will copy the setting from legacy to app specific
+			// For 4, we will do nothing.
+
+			// History:
+			// When persistent was implemented, it relied on saving the PrimaryLanguageOverride value to
+			// "__Uno.PrimaryLanguageOverride" in LocalSettings of ApplicationData.
+			// After that, we found that in Skia, the LocalSettings of ApplicationData is shared across all applications (which is wrong).
+			// The fix for having app-specific local settings on Skia is being done for Uno 5 (see https://github.com/unoplatform/uno/pull/12314)
+			// However, since the above is breaking change, we won't take it for Uno 4.x, but we still want PrimaryLanguageOverride to behave properly in Uno 4.x.
+			// So, we modify the key to have an app-specific part.
+			// We try to copy the value we read from the "legacy" key to the app-specific key.
+			// In a future major release, we may consider removing the legacy setting key.
+
+			// Note: There is a Wasm concern that is mentioned in https://github.com/unoplatform/uno/issues/12320
+			// So, we decided to implement this for all platforms.
+
+			if (hasAppSpecificSetting)
 			{
-				_primaryLanguageOverride = stringSavedValue;
+				_primaryLanguageOverride = appSpecificSavedValue;
+			}
+			else if (hasLegacySetting)
+			{
+				ApplicationData.Current.LocalSettings.Values[appSpecificKey] = legacySavedValue;
+				_primaryLanguageOverride = legacySavedValue;
 			}
 #endif
 
 			ApplyLanguages();
 		}
 
+		private static bool TryGetSettingFromKey(string key, out string value)
+		{
+			if (ApplicationData.Current.LocalSettings.Values.TryGetValue(key, out var savedValue))
+			{
+				value = savedValue as string;
+				return value is not null;
+			}
+
+			value = null;
+			return false;
+		}
+
+		private static string GetAppSpecificSettingKey()
+		{
+			return $"PrimaryLanguageOverrideSettingKey.{Package.EntryAssembly.GetName().Name}";
```

Comment:
If someone tried to access ApplicationLanguages too early, before an Application instance is created. This is going to blow up.

@MartinZikmund Thoughts? Is it a valid use case to access ApplicationLanguages before Application instance is created? Not sure how to best approach this.

Reply:
Technically it should never happen - I imagine the earliest usage in the `App` constructor, which is after `Current` was already set in the `Application` base constructor. I would suggest throwing a clear exception here (something like you are trying to use `ApplicationLanguages` too early in tyhe application lifecycle).

Reply:
@MartinZikmund One could attempt to access it before this line, for example:

https://github.com/unoplatform/uno.templates/blob/de24edc2b0f8cbbd14a988fb5d93ed21b0a23288/src/Uno.Templates/content/unoapp/MyExtensionsApp._1.Skia.Gtk/Program.cs#L19

I'll throw an exception.

Reply:
Actually, based on my tests, it is possible to use even the `Package.Current` API in WinUI before the app starts:

```csharp
static void Main(string[] args)
{
    XamlCheckProcessRequirements();

    global::WinRT.ComWrappersSupport.InitializeComWrappers();
    var name = Package.Current.Id.Name;
    global::Microsoft.UI.Xaml.Application.Start((p) => {
        var context = new global::Microsoft.UI.Dispatching.DispatcherQueueSynchronizationContext(global::Microsoft.UI.Dispatching.DispatcherQueue.GetForCurrentThread());
        global::System.Threading.SynchronizationContext.SetSynchronizationContext(context);
        new App();
    });
}
```

So I would vote for making the `SetEntryAssembly` part of the host initialization sequence somehow. Let's discuss later

Reply:
```
Failed: VerifyMultipleEraCalendar() [System.TypeInitializationException] 
 System.TypeInitializationException: TypeInitialization_Type, Windows.UI.Xaml.Controls.CalendarView
 ---> System.TypeInitializationException: TypeInitialization_Type, Windows.Globalization.ApplicationLanguages
 ---> System.InvalidOperationException: ApplicationLanguages is being accessed too early before an instance of Application was created.
   at Windows.Globalization.ApplicationLanguages.GetAppSpecificSettingKey()
   at Windows.Globalization.ApplicationLanguages..cctor()
   Exception_EndOfInnerExceptionStack
   at Windows.Globalization.Calendar..ctor()
   at Windows.UI.Xaml.Controls.CalendarView.GetOrCreateGregorianCalendar()
   at Windows.UI.Xaml.Controls.CalendarView.GetDefaultMaxDate()
   at Windows.UI.Xaml.Controls.CalendarView..cctor()
   Exception_EndOfInnerExceptionStack
   at Private.Infrastructure.CalendarHelper.CalendarViewHelper.<GetCalendarView>b__1_0()
   at Private.Infrastructure.TestServices.RunOnUIThread(Action action)
   at Private.Infrastructure.CalendarHelper.CalendarViewHelper.GetCalendarView()
   at Windows.UI.Xaml.Tests.Enterprise.CalendarViewIntegrationTests.VerifyMultipleEraCalendar()
   at Private.Infrastructure.CalendarHelper.CalendarViewHelper.<GetCalendarView>b__1_0()
   at Private.Infrastructure.TestServices.RunOnUIThread(Action action)
   at Private.Infrastructure.CalendarHelper.CalendarViewHelper.GetCalendarView()
   at Windows.UI.Xaml.Tests.Enterprise.CalendarViewIntegrationTests.VerifyMultipleEraCalendar()
   at Private.Infrastructure.CalendarHelper.CalendarViewHelper.<GetCalendarView>b__1_0()
   at Private.Infrastructure.TestServices.RunOnUIThread(Action action)
   at Private.Infrastructure.CalendarHelper.CalendarViewHelper.GetCalendarView()
   at Windows.UI.Xaml.Tests.Enterprise.CalendarViewIntegrationTests.VerifyMultipleEraCalendar()
   at Uno.UI.Samples.Tests.UnitTestsControl.<>c__DisplayClass63_1.<<ExecuteTestsForInstance>g__InvokeTestMethod|2>d.MoveNext()
```

😕

Reply:
I wasn't able to figure out what is happening here. It is Wasm-specific